src/thread.cpp

```diff
@@ -0,0 +1,184 @@
+#include <time.h>
+#include "os.h"
+#include "thread.h"
+
+pthread_key_t ProfiledThread::_tls_key;
+int ProfiledThread::_buffer_size = 0;
+std::atomic<int> ProfiledThread::_running_buffer_pos(0);
+std::vector<ProfiledThread*> ProfiledThread::_buffer(0);
+
+void ProfiledThread::initTLSKey() {
+    static pthread_once_t tls_initialized = PTHREAD_ONCE_INIT;
+    pthread_once(&tls_initialized, doInitTLSKey);
+}
+
+void ProfiledThread::doInitTLSKey() {
+    pthread_key_create(&_tls_key, freeKey);
+}
+
+inline void ProfiledThread::freeKey(void *key) {
+    ProfiledThread* tls_ref = (ProfiledThread*)(key);
+    if (tls_ref != NULL) {
+        delete tls_ref;
+    }
+}
+
+void ProfiledThread::initCurrentThread() {
+    initTLSKey();
+    
+    ProfiledThread* tls = (ProfiledThread*)pthread_getspecific(_tls_key);
+    if (tls == NULL) {
+        int tid = OS::threadId();
+        tls = ProfiledThread::forTid(tid);
+        pthread_setspecific(_tls_key, (const void*) tls);
+    }
+}
+
+void ProfiledThread::initExistingThreads() {
+    static pthread_once_t initialized = PTHREAD_ONCE_INIT;
+    pthread_once(&initialized, doInitExistingThreads);
+}
+
+void ProfiledThread::initCurrentThreadWithBuffer() {
+    initTLSKey();
+
+    if (pthread_getspecific(_tls_key) != NULL) {
+        // if there is already a TLS value associated just bail out
+        return;
+    }
+
+    ProfiledThread* tls_ref = NULL;
+    int pos = _running_buffer_pos++;
+    if (pos < _buffer_size) {
+        tls_ref = _buffer[pos];
+        tls_ref->_tid = OS::threadId();
+    }
+    if (tls_ref != NULL) {
+        pthread_setspecific(_tls_key, (const void*)tls_ref);
+    } else {
+        const char* msg = "ProfiledThread TLS buffer too small.";
+        Profiler::instance()->writeLog(LOG_WARN, msg, strlen(msg));
+    }
+}
+
+void* ProfiledThread::delayedUninstallUSR1(void* unused) {
+    initTLSKey();
+
+    int res = 0;
+    struct timespec ts;
+    ts.tv_sec = 0;
+    ts.tv_nsec = 1000000;
+    // wait for the TLS to be set
+    while (!res && errno != EINTR && pthread_getspecific(_tls_key) == NULL) {
```

Comment:
I am not sure to understand the loop condition: you should not check `errno` when `res==0`, `errno` is only set when `res!=0`.
If you want to restart sleep after EINTR, the loop should be:
```suggestion
    while ((!res || errno == EINTR) && pthread_getspecific(_tls_key) == NULL) {
```